File: WoodPuzzle/src/woodPuzzle/model/Configuration.java
package woodPuzzle.model;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Configuration {
	
	private final Puzzle puzzle;
	private Shape cells[];
	private Set<Shape> unusedShapes; 

	public Shape[] getCells() {
		return cells;
	}
	
	public Set<Shape> getUnusedShapes() {
		return unusedShapes;
	}

	/**
	 * The base constructor. Creates an empty configuration with
	 * the given puzzle state.
	 * @param p A reference to the puzzle description.
	 */
	public Configuration(Puzzle p) {
		this.puzzle = p;
		this.cells = new Shape[p.totalCells]; 
		this.unusedShapes = new HashSet<Shape>(p.shapes);
		this.init();
	}
	
	/**
	 * The copy constructor. Creates a configuration that is a 
	 * copy of the given configuration that is entirely state
	 * independent of the given configuration, i.e. changes to this
	 * configuration will not affect the state of the given configuration.
	 * @param c A reference to the configuration to copy
	 */
	public Configuration(Configuration c) {
		this.puzzle = c.puzzle;
		this.cells = new Shape[c.cells.length];
		System.arraycopy(c.cells, 0, this.cells, 0, this.cells.length);
		this.unusedShapes = new HashSet<Shape>(c.unusedShapes);
	}
	
	/**
	 * Initialized the cells of this configuration by setting all
	 * references to null. This should not be used with the copy constructor. 
	 */
	private void init() {
		for (int i = 0; i < puzzle.getTotalCells(); i++) {
			this.cells[i] = null;
		}
	}
	
	/**
	 * Attempts to place the given shape in the position specified
	 * by the given list of coordinates. The following must be true:
	 *  1) The given coordinates are valid
	 *  2) The given coordinates do not overlap with a placed shape
	 *  3) The list of coordinates correspond to the given shape
	 * If the shape is placed successfully, it is removed from the set
	 * of unused shapes and the given coordinates in the cells of this
	 * configuration will correspond to the given shape.
	 * @param shape A reference to the shape to place.
	 * @param position The list of coordinates to place the shape.
	 * @return true if the above conditions are true, otherwise false.
	 */
	public boolean placeShape(Shape shape, List<Coordinate> position) {
		if (!unusedShapes.contains(shape)) return false;
		int lowx = Integer.MAX_VALUE, lowy = Integer.MAX_VALUE, lowz = Integer.MAX_VALUE;
		for (Coordinate c : position) {
			if (!puzzle.isValidCoordinate(c)) return false;
			if (cells[puzzle.hashCoordinate(c)] != null) return false;
			if (c.x < lowx) lowx = c.x;
			if (c.y < lowy) lowy = c.y;
			if (c.z < lowz) lowz = c.z;
		}
		int shapeTotal = puzzle.shapeSide * puzzle.shapeSide * puzzle.shapeSide;
		int temp[] = new int[shapeTotal];
		for (int i = 0; i < shapeTotal; i++) temp[i] = 0;
		for (Coordinate c : position) {
			temp[shape.hashCoordinate(c.vectorAdd(-lowx, -lowy, -lowz))] = 1;
		}
		
		if (!isIdenticalRotatedShape(shape, temp)) return false;
		
		for (Coordinate c : position) {
			cells[puzzle.hashCoordinate(c)] = shape;
		}
		
		unusedShapes.remove(shape);
		
		return true;
	}
	
	/**
	 * Checks whether there exists a rotation of the first shape 
	 * that is identical to the second shape. It is assumed that s2  
	 * is already placed "against" the origin. 
	 * @param s1 A reference to the first shape
	 * @param s2 The cells of the second shape
	 * @return true if such a rotation exists, false otherwise.
	 */
	private boolean isIdenticalRotatedShape(Shape s1, int[] s2) {
		for (int yaxis = 0; yaxis <= 3; yaxis++) {
			for (int zaxis = 0; zaxis <= 3; zaxis++) {
				int[] temp;
				temp = s1.rotateShape(yaxis, zaxis);
				if (isIdenticalShape(temp, s2)) return true;
			}
		}
		
		return false;
	}
	
	/**
	 * Checks whether the two given shapes are perfectly identical, 
	 * specifically there is no check for positioning or rotation. It is
	 * assumed that s1 and s2 are already be placed "against" the origin.
	 * @param s1 The cells of a first shape
	 * @param s2 The cells of a second shape
	 * @return true if they are perfectly identical, otherwise false.
	 */
	private boolean isIdenticalShape(int[] s1, int[] s2) {
		for (int i = 0; i < s1.length; i++) {
			if (s1[i] != s2[i]) return false;			
		}
		return true;
	}
	
	/**
	 * Removes a shape from the set of unused shapes (discards it)
	 * @param s A reference to the shape to remove
	 */
	public void removeShape(Shape s) {
		if (unusedShapes.contains(s)) unusedShapes.remove(s);
	}
}

File: src/woodpuzzle/model/Configuration.kt
package woodpuzzle.model

data class Configuration(
    val puzzle: Puzzle,
    val cells: Array<Shape?>,
    val unusedShapes: Set<Shape>
) {

    /**
     * The base constructor. Creates an empty configuration with
     * the given puzzle state.
     * @param puzzle A reference to the puzzle description.
     */
    constructor(puzzle: Puzzle) : this(
        puzzle = puzzle,
        cells = arrayOfNulls(puzzle.totalCells),
        unusedShapes = puzzle.shapes
    )

    fun allCellsFilled(): Boolean = !cells.contains(null)

    /**
     * Attempts to place the given shape in the position specified
     * by the given list of coordinates. The following must be true:
     * 1) The given coordinates are inside the box
     * 2) The given coordinates do not collide with a placed shape
     * If the shape is placed successfully, a new configuration object is
     * created without the placed shape in the set of unused shapes and
     * the given coordinates in the cells of the new configuration will
     * point to the given shape.
     * @param shape A reference to the shape to place.
     * @param placement The list of coordinates to place the shape.
     * @return a new configuration if the above conditions are true, otherwise null.
     */
    fun placeShape(shape: Shape, placement: List<Coordinate>): Configuration? {
        // Check whether the reference is even an unused shape.
        // Should never be false but don't really trust myself!
        if (!unusedShapes.contains(shape)) return null
        for (coordinate in placement) {
            // Check whether the coordinate is still in the bounds of the puzzle box.
            if (!puzzle.isValidCoordinate(coordinate)) return null
            // Check whether this is a collision with an existing placement in the puzzle.
            if (cells[puzzle.hashCoordinate(coordinate)] != null) return null
        }
        // At this point we know the shape can be placed. Create a copy of this configuration.
        val newConfig = this.copy(cells = this.cells.copyOf(), unusedShapes = this.unusedShapes - shape)
        // Place the shape and remove it from unused in the new configuration.
        for (c in placement) newConfig.cells[puzzle.hashCoordinate(c)] = shape
        return newConfig
    }

    /**
     * Checks whether this configuration has dead cells. That is,
     * if a group of empty and connected cells is smaller than the
     * given minimum shape size, then it is dead. Furthermore,
     * if all shapes are of identical size then a similar group with
     * the number of empty cells not a multiple of the shape size is
     * also considered dead.
     * @return true if there are dead cells, otherwise false.
     */
    fun hasDeadCells(): Boolean {
        val visited = BooleanArray(puzzle.totalCells) // Inits to false
        for (x in 0 until puzzle.width) {
            for (y in 0 until puzzle.height) {
                for (z in 0 until puzzle.length) {
                    // Iterates over every cell of the box looking for an unvisited, empty gap.
                    val currentIndex = puzzle.hashCoordinate(x, y, z)
                    if (visited[currentIndex]) continue
                    visited[currentIndex] = true
                    if (cells[currentIndex] != null) continue
                    // If we are here then the current cell is unvisited and an empty gap, so
                    // count and visit the whole gap.
                    var emptyCount = 1
                    val checkNeighbours: MutableList<Coordinate> = mutableListOf(Coordinate(x, y, z))
                    while (checkNeighbours.isNotEmpty()) {
                        val c = checkNeighbours.removeFirst()
                        val coordinatesToVisit: List<Coordinate> = listOf(
                            Coordinate(c.x + 1, c.y, c.z),
                            Coordinate(c.x - 1, c.y, c.z),
                            Coordinate(c.x, c.y + 1, c.z),
                            Coordinate(c.x, c.y - 1, c.z),
                            Coordinate(c.x, c.y, c.z + 1),
                            Coordinate(c.x, c.y, c.z - 1),
                        )
                        for (adjacentCoordinate in coordinatesToVisit) {
                            if (!puzzle.isValidCoordinate(adjacentCoordinate)) continue
                            val adjacentIndex = puzzle.hashCoordinate(adjacentCoordinate)
                            if (visited[adjacentIndex]) continue
                            visited[adjacentIndex] = true
                            if (cells[adjacentIndex] == null) {
                                emptyCount++
                                checkNeighbours.add(adjacentCoordinate)
                            }
                        }
                    }
                    // If a gap is smaller than the smallest possible shape, it's dead
                    if (emptyCount < puzzle.minShapeSize) return true
                    // If the shapes are all the same size, and the gap is not a multiple
                    // of the shape size, then it's also dead
                    if (puzzle.minShapeSize == puzzle.maxShapeSize &&
                        emptyCount % puzzle.maxShapeSize != 0) return true
                }
            }
        }
        return false
    }
}
